### PR TITLE
http2: Remove RELEASE_ASSERTs in sendPendingFrames() error handling

### LINKED ISSUE #12280: [HTTP/2] implement protocol error handling out of dispatch context
HTTP/2 codec uses private **sendPendingFrames()** for building outbound wire bytes. This method may fail if higher level protocol constraints are violated. It is also possible that **nghttp2** methods for building frames fail. The high level protocol constraints checked by the codec, consist of limits on the number of outbound frames of different types, and only implemented for server codecs at this point.
There are presently 11 call sites of this method in the codec. Only the call site in the **dispatch()** method checks and handles the error code returned (or thrown in the legacy implementation) by the sendPendingFrames() method. The other 10 call sites cause abnormal program termination if the **sendPendingFrames()** encounters an error.
The error handling is implemented by adding a variable tracking (latching) the error state of the codec and a closure that is executed when codec is transitioned into the error state. The closure is executed by the event dispatcher after all events in the current event loop iteration have been processed.
The error state has to be tracked by codec to handle the case where an I/O event is pending for failed codec in the current event loop iteration. The error state is checked before the **dispatch()** method is invoked and if the codec is in the failed state it is torn down instead of processing inbound data.
Error handling will be added to each **sendPendingFrames()**call site one by one to minimize the risk and amount of changes needed to be reviewed in each PR. When the **sendPendingFrames()** method returns an error code (or throws an exception in the legacy implementation) the error state of codec is changed to failed and the closure for tearing down the codec is scheduled to run by the event dispatcher at the end of the current iteration.

Depends on #12279
Sub-task of #12278

### PR DESCRIPTION
**Pull request by yanavlasov**
Commit Message:
Remove RELEASE_ASSERTs in sendPendingFrames() error handling

Additional Description:
This PR removes RELEASE_ASSERT on the status of the `sendPendingFrames()` call and replaces it with the error handling that schedules callback to disconnect connection. The change makes `sendPendingFrames()` for server codecs to always return the status of the protocol constraint check regardless of the dispatching/non-dispatching context.

During tests I have discovered that returning an error from the `onSend` and `onDataSourceSend` callbacks in non-dispatching context does not work well because it aborts the processing of the `nghttp2_session_send()` and as a result any nghttp2 callbacks that may have been invoked on pending frames will not be called. This is a problem for frames like RST_STREAM and GOAWAY because they invoke callbacks in codecs and http connection manager.

As a result I have unwound the code that returns errors from the `onSend` and `onDataSourceSend` callbacks when outbound frame queue limits had been exceeded and instead let the `nghttp2_session_send()` complete normally and check for protocol constraint violations after it has completed.

Risk Level: Medium
Testing: Unit Tests
Docs Changes: N/A
Release Notes: N/A
Fixes #12280

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

File: source/common/http/http2/codec_impl.cc [hunk 1]
```diff
@@ -416,18 +397,17 @@ Status ConnectionImpl::StreamImpl::onDataSourceSend(const uint8_t* framehd, size
   parent_.protocol_constraints_.incrementOutboundDataFrameCount();
 
   Buffer::OwnedImpl output;
-  auto status = parent_.addOutboundFrameFragment(output, framehd, FRAME_HEADER_SIZE);
-  if (!status.ok()) {
+  parent_.addOutboundFrameFragment(output, framehd, FRAME_HEADER_SIZE);
+  if (!parent_.protocol_constraints_.checkOutboundFrameLimits().ok()) {
     ENVOY_CONN_LOG(debug, "error sending data frame: Too many frames in the outbound queue",
                    parent_.connection_);
     setDetails(Http2ResponseCodeDetails::get().outbound_frame_flood);
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:404]: Should there be an error return for this case, or should this be void and the call site always return 0?

File: source/common/http/http2/codec_impl.cc [hunk 2]
```diff
@@ -490,10 +467,7 @@ void ConnectionImpl::StreamImpl::encodeDataHelper(Buffer::Instance& data, bool e
     data_deferred_ = false;
   }
 
-  auto status = parent_.sendPendingFrames();
-  // See comment in the `encodeHeadersBase()` method about this RELEASE_ASSERT.
-  RELEASE_ASSERT(status.ok(), "sendPendingFrames() failure in non dispatching context");
-  parent_.checkProtocolConstraintViolation();
+  parent_.sendPendingFramesAndHandleError();
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:470]: question before I review farther, does it make sense to continue processing if an error handler has been scheduled? Most of these are the last function called.

Also, where we sendPendingFramesAndHandleError, are we confident we have tests for error handling at each call site?  Would it make sense to always 
if (sendPendingFramesAndHandleError()) {
  return;
}
so we can coverage-verify the error handling is tested?  Or given these replace release asserts are we asserting they can't happen and mainly hoping they don't cause issues?

**Review by alyssawilk — COMMENTED** [review]
Wohoo!  I'm all about reduced release asserts :-)

File: source/common/http/http2/codec_impl.cc (hunk 1 above)
Comment (yanavlasov) [reply source/common/http/http2/codec_impl.cc:416]: The error will be handled in the `sendPendingFrames()`. I've made this function void.

File: source/common/http/http2/codec_impl.cc (hunk 2 above)
Comment (yanavlasov) [reply source/common/http/http2/codec_impl.cc]: Yes, I think in some cases additional processing, like scheduling timers is not needed. I have changed these paths.
I have added in last few PRs 16 new integration test cases for all invocations of the `sendPendingFrames()`. To the best of my analysis all call sites in all possible contexts are covered. It does not mean that I'm right of course. So I've added the `if` clause as you have suggested and will check the coverage output.

**Review by mattklein123 — APPROVED** [review]
Awesome work!!! LGTM. Code is much easier to understand now. Will defer to @alyssawilk and others for any final comments.

**Review by alyssawilk — APPROVED** [review]
Awesome, thanks!

File: source/common/http/http2/codec_impl.cc [hunk 3]
```diff
@@ -1120,7 +1090,23 @@ Status ConnectionImpl::sendPendingFrames() {
...
+    ENVOY_CONN_LOG(debug, "error sending frames: Too many frames in the outbound queue.",
+                   connection_);
+  }
+  return status;
+}
+
+bool ConnectionImpl::sendPendingFramesAndHandleError() {
+  if (!sendPendingFrames().ok()) {
+    scheduleProtocolConstraintViolationCallback();
+    return true;
```
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.cc:1107]: returning true on error is a bit unusual, but you are documenting the return value so it seems fine.

**Review by antoniovicente — APPROVED** [review]
Very nice change.  Go go improvements to H2 codec safety.

**Comment by mattklein123** [comment]
I think CI issues should be fixed. Let's try one more main merge?

/wait

**Comment by mattklein123** [comment]
Merge main once https://github.com/envoyproxy/envoy/pull/13598 merges. Thanks!

/wait

**Pull request merged**